Freelygive/ai recipe

ivanboring opened the pull request:
## Linked issues

- #10 

## Solution

- Adds the [AI](https://www.drupal.org/project/ai) and [AI Image Al Text](https://www.drupal.org/project/ai_image_alt_text) module into the recipe.
- Enables [Key](https://www.drupal.org/project/key), AI, The OpenAI and Anthropic provider and the AI Image AI Text module.
- Fixes the AI Image Alt Text settings to autogenerate images and hide the button.
- Adds the choice of setting up an AI provider with its API key in the ConfigureAPIKeysForm.php.
- If a provider is choosen, it creates a Key entity and connects it to the provider configuration and sets that configuration as the default Chat provider. This has been tested to work without breaking `ddev quick-start`.

## Possible Todos
- Assign the `generate ai alt tags` permission to the correct roles.

## Expected Failures

**In Editor BS3**
``` shell
```

**In Anonymous BS3**
``` shell
```

**In Editor BS5**
``` shell
```

**In Anonymous BS5**
``` shell
```

## QA Tests

**QA Tests Changes**
<!-- 
List the test names as a list (e.g)

- Shared(BS3, BS5):
  - `TestName1`
  - `TestName2`
- BS3:
  - `TestName1`
  - `TestName2`
- BS5:
  - `TestName1`
  - `TestName2`

-->
## CI/CD

To run the QA tests write a new comment as follows:
- BS3
`/qa-demo-2x-bs3-tests`
- BS5
`/qa-demo-2x-bs5-tests`

## Checklist

<!--- Put an `x` in all the boxes that apply: -->
- [ ] I have read the [CONTRIBUTING.md](https://github.com/dxpr/dxpr_builder/blob/2.x/CONTRIBUTING.md) document.
- [ ] My commit messages follow the contributing standards and style of this project.
- [ ] My code follows the coding standards and style of this project.
- [ ] My code contains no changes that are outside the scope of the issue.
- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [ ] Need to run update.php after code changes.
- [ ] Requires a change to end-user documentation.
- [ ] Requires a change to developer documentation.
- [ ] Requires a change to QA tests.
- [ ] Requires a new QA test.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.








<!-- This is an auto-generated comment: release notes by OSS CodeRabbit -->
### Summary by CodeRabbit

- New Feature: Added support for OpenAI and Anthropic AI providers in the configuration settings.
- New Feature: Introduced API key validation for both OpenAI and Anthropic providers to ensure correct setup.
- Refactor: Implemented a function to uninstall unused AI provider modules, enhancing system efficiency by removing unnecessary components. 
- Chore: Updated the `ConfigureAPIKeysForm` class to handle new AI provider settings and API keys management.
<!-- end of auto-generated comment: release notes by OSS CodeRabbit -->

ivanboring commented:
@jjroelofs - please feel free to assign to the reviewers that should check it, if its not you.

Note the comment about permissions, if its know which roles should have the access, I can add this as well.

jjroelofs commented:
@ivanboring what comment about permissions?

ivanboring commented:
@jjroelofs Sorry, it was under possible todo's, so:
`Assign the generate ai alt tags permission to the correct roles.`

Should it be assigned to content_editor, content_administrator, marketer, site_bui etc?

jjroelofs commented:
@ivanboring right, I think we can assign it to all roles, since adding, editing, or managing media is relevant in some way to all of them.

jjroelofs commented on installer/src/Form/ConfigureAPIKeysForm.php:
```diff
@@ -164,7 +223,83 @@ public function validateForm(array &$form, FormStateInterface $form_state): void
+   */
+  protected function testOpenAIKey(string $openai_key): string {
+    $client = OpenAI::client($openai_key);
+    try {
+      $client->models()->list();
+    }
+    catch (\Exception $e) {
+      return $e->getMessage();
+    }
+    return '';
   }
 
+  /**
+   * Simple test of the Anthropic Key.
+   *
+   * @param string $anthropic_key
+   *   The Anthropic key.
+   *
+   * @return string
+   *   If the key is valid, the message.
+   */
+  protected function testAnthropicKey(string $anthropic_key): string {
+    $client = new AnthropicAPI($anthropic_key);
+    $payload = [
+      'model' => 'claude-3-5-sonnet-20240620',
+      'messages' => [
+        [
+          'role' => 'user',
+          'content' => 'Say hello!',
+        ],
+      ],
+    ];
+    try {
+      $client->messages()->maxTokens(10)->create($payload);
+    }
+    catch (\Exception $e) {
+      return $e->getMessage();
+    }
+    return "";
+  }
```
@ivanboring is it possible we
* Apply some DRY as suggested by automatic review
* Not hardcode a model ID as this model might be deprecated in x years and we will not find out easily that this code is not working, as entering an Anthropic key has no test coverage. Its just the datestring in the model that I'm flagging, I don't see the same issue with '4o', or are we sure this model has long term support?

jjroelofs commented on recipes/dxpr_cms_image_media_type/recipe.yml:
```diff
@@ -6,9 +6,15 @@ recipes:
+  - key
+  - ai
+  - provider_openai
+  - provider_anthropic
```
I can see how we need to enable this for the key check during installation, but I'm inclined to think would be better if we:
1. Enable them programmatically before the Keys profile page loads
2. Disable whichever is/are not used in the cleanup step:
 https://github.com/dxpr/dxpr_cms/blob/e9b9e5f6146cb568c7f9d13a28a90922a9c7ba6c/installer/dxpr_cms_installer.profile#L159-L166

I might be wrong, let me know your thoughts

ivanboring replied:
It would make sense to uninstall the not used one as a last step, because then I could use the actual installed provider for doing the authentication check and get a list of available models in the provider. The AI module has the same problem, that we need to hardcode this date as a config, because it's not provided via the API from Anthropic, but at least we will keep an eye out for if it changes (and I'll add a [test](https://www.drupal.org/project/ai/issues/3469650) for it)

ivanboring replied:
@ivanboring  This is updated now, with the above mentioned updates. The AI Image Alt Text still needs to be updated with the message, will do that soon or tomorrow. There are two notes:

1. If No AI is chosen, it still keeps both providers installed, it only uninstalls the ones who are not being used, if one is installed. Correct?
2. The recipe that creates the roles has the media recipe as dependency, which means that the permission can't be set without causing circular dependencies. I've added so it gets created if its missing, which solves it for DXPR, but it also means that anyone that installs the recipe outside of DXPR, gets these roles. Correct?

jjroelofs replied:
yes

jjroelofs reviewed (changes requested):
## UX Feedback

When installing with DDEV quick-start, the module is enabled but not working, and there is not feedback/warning as to why it is not working:
![image](https://github.com/user-attachments/assets/e8a76620-6ea4-416e-8ee8-416d21b5b729)

Would be great if we could print a warning to users with permissions for site administration with a link to the form for setting up a key and choosing a provider?

Maybe a second warning here would also be in order as setting a provider alone is not enough?

![image](https://github.com/user-attachments/assets/fd198927-7ff2-472d-ba39-4cbe931bd53a)


I haven't looked at the source code of the AI Alt text module yet so I hope you don't mind posting this feedback here.

jjroelofs commented:
@ivanboring on this page it also looks like nothing is wrong even though there is no key or provider configured after ddev quick-start:
![image](https://github.com/user-attachments/assets/a2c92b4b-d7a9-43ae-9980-1f7d8183c2e7)

jjroelofs commented:
@ivanboring after I add a key and set up the OpenAI provider, the image alt text module is still not working. Also, there is no UI in the Alt text module config form to choose a provider? Am I missing something?

ivanboring commented:
@jjroelofs - currently I'm waiting for[ this pull request](https://git.drupalcode.org/project/ai/-/merge_requests/45) on the AI module to be approved, hopefully today, when that is done there is already a branch in the AI Image Alt Text module that will fix the models being choosable specifically for the AI Image Al Text module in the settings.

Currently it is using the Default Chat AI provider that can be setup under /admin/config/ai/settings.

I will add a warning message, when its not installed (and maybe a log?) that tells you why its disabled.

jjroelofs commented:
@ivanboring heads up: there is a merge conflict because @Neslee implemented changes resulting from the Drupal Starshot > Drupal CMS migration (they copy our name :p). The changes are massive, and @Neslee will push to this branch shortly to get it sorted for you.

Neslee commented:
@jjroelofs @ivanboring sorry about it, i have resolved conflicts now.

ivanboring commented:
Now the final changes to the AI Image Alt Text module are done as well, you can see how the warning message would look here:
![no-provider-alt-text](https://github.com/user-attachments/assets/ad18afaf-35e1-419d-91ff-3d2493bcd958)

You should also be able to switch to another model directly in the settings form of the AI Image Alt Text settings.

This also updates the configuration form installer to set a default chat_with_image_vision, but also a default chat provider since its being used for many of the "normal" third party modules, like CKEditor, Translate etc.

ivanboring commented on dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php:
```diff
@@ -164,7 +246,80 @@ public function validateForm(array &$form, FormStateInterface $form_state): void
+    // @todo Change to chat_with_image_vision when it is available.
+    $models = $provider->getConfiguredModels('chat');
+    return key($models);
+  }
```
This makes sense, I have also reafctored so the actual provider form check is independent on provider.

ivanboring commented on dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php:
```diff
@@ -145,6 +197,36 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
     if (!empty($google_translation_key)) {
       $this->configFactory->getEditable('tmgmt.translator.google')->set('settings.api_key', $google_translation_key)->save();
     }
+
+    // If the AI provider is set, enable the appropriate modules.
+    if ($ai_provider = $form_state->getValue('ai_provider')) {
+      try {
+        // Setup the key for the provider.
+        $key_id = $ai_provider . '_key';
+        $key = Key::create([
+          'id' => $key_id,
+          'label' => ucfirst($ai_provider) . ' API Key',
+          'description' => 'API Key for ' . ucfirst($ai_provider),
+          'key_type' => 'authentication',
+          'key_provider' => 'config',
+        ]);
+        $key->setKeyValue($form_state->getValue($key_id));
+        $key->save();
+        // Add the key to the config.
+        $this->configFactory->getEditable('provider_' . $ai_provider . '.settings')->set('api_key', $key_id)->save();
+        // Set the default chat and chat_with_image_vision provider.
+        $this->configFactory->getEditable('ai.settings')->set('default_providers.chat', [
+          'provider_id' => $ai_provider,
+          'model_id' => $ai_provider == 'openai' ? 'gpt-4o' : $this->getFirstAiModelId($ai_provider),
+        ])->set('default_providers.chat_with_image_vision', [
+          'provider_id' => $ai_provider,
+          'model_id' => $ai_provider == 'openai' ? 'gpt-4o' : $this->getFirstAiModelId($ai_provider),
+        ])->save();
+      }
+      catch (\Exception $e) {
+        $this->messenger()->addError($this->t('An error occurred while saving the AI provider key: @error', ['@error' => $e->getMessage()]));
+      }
+    }
```
This key should not exist before, since its part of the installation form.

ivanboring commented on dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php:
```diff
@@ -119,6 +138,39 @@ public function buildForm(array $form, FormStateInterface $form_state, array &$i
       ];
     // }
 
+    $form['ai_provider'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Select AI Provider'),
+      '#description' => $this->t('If you want to enable AI features like ai powered alt text generation, select the AI provider you want to use for AI features and fill in the API Key.'),
+      '#empty_option' => $this->t('No AI'),
+      '#options' => [
+        'openai' => $this->t('OpenAI'),
+        'anthropic' => $this->t('Anthropic'),
+      ],
+    ];
+
+    $form['openai_key'] = [
+      '#type' => 'password',
+      '#title' => $this->t('OpenAI API key'),
+      '#description' => $this->t('Get a key from <a href="https://platform.openai.com/api-keys" target="_blank">platform.openai.com/api-keys</a>.'),
+      '#states' => [
+        'visible' => [
+          ':input[name="ai_provider"]' => ['value' => 'openai'],
+        ],
+      ],
+    ];
+
+    $form['anthropic_key'] = [
+      '#type' => 'password',
+      '#title' => $this->t('Anthropic API key'),
+      '#description' => $this->t('Get a key from <a href="https://console.anthropic.com/settings/keys" target="_blank">console.anthropic.com/settings/keys</a>.'),
+      '#states' => [
+        'visible' => [
+          ':input[name="ai_provider"]' => ['value' => 'anthropic'],
+        ],
+      ],
+    ];
+
```
This follows how its done with Google Translate API Key.

ivanboring commented on dxpr_cms_installer/dxpr_cms_installer.profile:
```diff
@@ -21,6 +21,9 @@ function dxpr_cms_installer_install_tasks(): array {
        'type' => 'form',
        'function' => ConfigureAPIKeysForm::class,
     ],
+    'dxpr_cms_uninstall_unused_ai_modules' => [
+      // Uninstall the unused AI provider module.
+    ],
```
Following the code written before.

jjroelofs reviewed (changes requested):
Great work, normally you would re-request my review until my approval but since I'm not available until Wednesday, after fixing the below issues please request @Neslee's review right away to save time.

1. In the help message, please link to the respective config pages (messages should help the user get to the place where they fix the problem)
2. The install wizard is not recognizing my OpenAI API key, even though it works fine to generate alt text if I install it after the wizard completes. I'll email you the key for testing
![image](https://github.com/user-attachments/assets/11cc8838-f045-41e0-ade2-9a8a074bf8db)

ivanboring commented:
@Neslee The OpenAI key should work, now - the organizational keys are 131 character long and the default password field was only 128 characters.

For the link in the message, this is in the open source module, but provided is a screenshot and the link to the changes are in this commit: https://git.drupalcode.org/project/ai_image_alt_text/-/commit/6792eabc93f20165cc17ffc14b97b6038c58ef4c

![with-links](https://github.com/user-attachments/assets/ac1f04be-5011-43ad-baa1-8e57c45e5d11)

Neslee commented:
@ivanboring thanks for the update. I installed my site using your branch and completed the setup in the Drupal installation wizard.

1. I included the correct OpenAI or Anthropic AI key, and it was successfully added to the configuration.
2. I tried entering an incorrect OpenAI key and received the following error message. It seems there is a typo with two dots. The same typo with two dots also appeared when I used an incorrect Anthropic AI key. It seems this issue is coming from the provider.
<img width="593" alt="Screenshot 2024-08-27 at 2 02 43 PM" src="https://github.com/user-attachments/assets/3c6eafef-8ab3-4402-a9ef-deb659961751">

3. Once the site was up and running, I navigated to /admin/config/ai/ai_image_alt_text and updated the configurations as shown in the screenshot below.
<img width="1333" alt="Screenshot 2024-08-27 at 2 20 08 PM" src="https://github.com/user-attachments/assets/7d17edd5-c2f6-494c-8b3b-619367042262">

Then I navigated to the node page and attempted to add media. After uploading the media, a throbber appeared and continued loading without successfully uploading the image, as shown in the screenshot below.
<img width="1470" alt="Screenshot 2024-08-27 at 2 11 32 PM" src="https://github.com/user-attachments/assets/cd4e5b79-7517-41da-8c2c-d322b4fc2e1f">

The same thing happens if I navigate to /media/add/image and try to add an image.
<img width="1470" alt="Screenshot 2024-08-27 at 2 12 07 PM" src="https://github.com/user-attachments/assets/dab4023d-7e3b-4c3f-acf7-bd6a6b7200a0">

4. When I change the AI provider/model to the default image vision model, I receive the following message while adding an image
<img width="1018" alt="Screenshot 2024-08-27 at 2 26 57 PM" src="https://github.com/user-attachments/assets/f6d15268-bda3-4504-8d30-b8bd3b197baa">

Please let me know if I've missed any settings or configurations. Also, there are new commits on the main branch; it would be good to merge them into this branch since there are some new patches that we need to run the recipe-based Drupal site.

ivanboring commented:
@Neslee:

2. The double should be gone now.
4. This error message was a bug, its fixed in the AI Image Alt Text module now.

Merge is also done now.

For #3, I could not replicate the issue at first, what I did was:
1. Installed the system without an AI provider.
2. Setup OpenAI.
3. Setup the AI Image Al Text according to your settings.
4. Upload an image.

This work, however I changed the OpenAI key to a key that doesn't work and I could replicate the issue. I hope that is the same issue as you had? I've added a fix to the AI Image Alt Text that takes care of this exception, currently it was just working on time outs. The system will give you a generic warning message now that it couldn't create the text and let you continue manually.

If this was the issue and you have the system still, you should be able to see in the logs an access denied message.

So 3 & 4 should work by using composer update/ddev rebuild on the AI Image Alt Text module and 2 and the merge, by pulling this repo.

Neslee commented:
@ivanboring  now I am getting the below warning message when I have the wrong API key, which works as expected
<img width="975" alt="Screenshot 2024-08-27 at 6 56 14 PM" src="https://github.com/user-attachments/assets/5479848e-d6c4-4645-b1aa-d9569090fc23">

Everything else looks good based on the configurations and code. Thanks. Let's wait for @jjroelofs review

jjroelofs reviewed (approved):
All blocking issues are solved now, thx! Not sure what happened with that key, I tried another key and it worked fine. Hope it's not a problem with legacy user keys versus project keys but that would not really be in the scope of this PR anyways.
